font/src/directwrite/mod.rs

```diff
@@ -186,29 +276,85 @@ pub enum Error {
     MissingFont(FontDesc),
     MissingGlyph(char),
     FontNotLoaded,
+    DirectWriteError(HRESULT),
 }
 
-impl ::std::error::Error for Error {
-    fn description(&self) -> &str {
-        match *self {
-            Error::MissingFont(ref _desc) => "Couldn't find the requested font",
-            Error::MissingGlyph(ref _c) => "Couldn't find the requested glyph",
-            Error::FontNotLoaded => "Tried to operate on font that hasn't been loaded",
-        }
-    }
-}
+impl std::error::Error for Error {}
 
-impl ::std::fmt::Display for Error {
-    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
-        match *self {
-            Error::MissingGlyph(ref c) => write!(f, "Glyph not found for char {:?}", c),
-            Error::MissingFont(ref desc) => write!(
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            Error::MissingGlyph(c) => write!(f, "Glyph not found for char {:?}", c),
+            Error::MissingFont(desc) => write!(
                 f,
                 "Couldn't find a font with {}\n\tPlease check the font config in your \
                  alacritty.yml.",
                 desc
             ),
             Error::FontNotLoaded => f.write_str("Tried to use a font that hasn't been loaded"),
+            Error::DirectWriteError(hresult) => {
+                write!(f, "A DirectWrite rendering error occurred: {:#X}", hresult)
+            }
+        }
+    }
+}
+
+fn em_size(size: Size) -> f32 {
+    size.as_f32_pts() * (96.0 / 72.0)
+}
+
+impl From<dwrote::Font> for Font {
+    fn from(font: dwrote::Font) -> Font {
+        Font {
+            face: font.create_font_face(),
+            family_name: font.family_name(),
+            weight: font.weight(),
+            style: font.style(),
+            stretch: font.stretch(),
+        }
+    }
+}
+
+impl From<Weight> for FontWeight {
+    fn from(weight: Weight) -> FontWeight {
+        match weight {
+            Weight::Bold => FontWeight::Bold,
+            Weight::Normal => FontWeight::Regular,
+        }
+    }
+}
+
+impl From<Slant> for FontStyle {
+    fn from(slant: Slant) -> FontStyle {
+        match slant {
+            Slant::Oblique => FontStyle::Oblique,
+            Slant::Italic => FontStyle::Italic,
+            Slant::Normal => FontStyle::Normal,
         }
     }
 }
+
+fn get_current_locale() -> String {
+    let mut buf = vec![0u16; LOCALE_NAME_MAX_LENGTH];
+    let len = unsafe { GetUserDefaultLocaleName(buf.as_mut_ptr(), buf.len() as i32) as usize };
+
+    // len includes null byte, we don't need that in Rust strings
+    OsString::from_wide(&buf[..len - 1]).into_string().expect("Locale not valid unicode")
+}
+
+/// Information needed by dwrote's TextAnalysisSource to find a fallback font
+/// for a run of text. See `get_fallback_font`.
```

Comment:
Usually the first line of documentation on an item should be a short description of what it is. I feel like if that doesn't even fit on a single line, it's probably too verbose.

```suggestion
/// Font fallback information for dwrote's TextAnalysisSource.
```

Do you think this correctly represents the purpose of this struct, or is there some information missing?